### PR TITLE
Build and push docker images using GitHub Actions

### LINKED ISSUE #245: Docker build should run in CI
CI should generate docker images and push them (for releases) to Docker Hub

### PR DESCRIPTION
**Pull request by reivilibre**
Closes #245.

### CODE REVIEW

**Comment by reivilibre** [comment]
https://hub.docker.com/repository/registry-1.docker.io/reivilibre/sygnal/tags?page=1&ordering=last_updated
https://github.com/reivilibre/sygnal

This seems to work

File: .github/workflows/docker_push.yml [hunk 1]
```diff
@@ -0,0 +1,64 @@
+# GitHub actions workflow which builds and publishes the docker images.
+
+name: Build and push docker images
+
+on:
+  push:
+    tags: ["v*"]
+    # uncomment this to enable all pushes to main
+    #branches: [ master, main ]
```
Comment (anoadragon453) [inline .github/workflows/docker_push.yml:9]: I think `latest` on our docker images always just points to the latest release. Not sure if we'd ever push on the `main` branch.

Though you mention re-enabling building docker images on main below - did we used to do that?

File: changelog.d/249.docker [hunk 2]
```diff
@@ -0,0 +1 @@
+Automatically build the Docker image and push it to Docker Hub using GitHub Actions.
```
Comment (anoadragon453) [inline changelog.d/249.docker:1]:
```suggestion
Automatically build the Docker image for each release and push it to Docker Hub using GitHub Actions.
```

File: .github/workflows/docker_push.yml [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+      - name: Log in to DockerHub
+        uses: docker/login-action@v1
+        with:
+          username: ${{ secrets.DOCKER_HUB_USERNAME }}
+          password: ${{ secrets.DOCKER_HUB_TOKEN }}
+
+      # The following line can be used to tag builds with the SHA hash of the commit.
+      # This would be useful if we wanted to re-enable building for all commits on `main`.
+      # type=sha,enable=true,priority=100,prefix=,suffix=,format=long
```
Comment (anoadragon453) [inline .github/workflows/docker_push.yml:40]: This is what would generate the following tag I assume?

image

**Review by anoadragon453 — COMMENTED** [review]
Looking great! Just a couple questions below.

File: .github/workflows/docker_push.yml (hunk 1 above)
Comment (reivilibre) [reply .github/workflows/docker_push.yml:9]: I ripped off https://github.com/matrix-org/synapse/blob/develop/.github/workflows/docker.yml which seems like if you push to master on Synapse, you will get a new build.

I'm not really sure why this is, so I took it out for Sygnal.

File: .github/workflows/docker_push.yml (hunk 3 above)
Comment (reivilibre) [reply .github/workflows/docker_push.yml:40]: Yeah, that's right. Reckon I should just cut it out entirely?

File: .github/workflows/docker_push.yml (hunk 1 above)
Comment (anoadragon453) [reply .github/workflows/docker_push.yml:9]: Given we don't have anything other than releases on [our dockerhub](https://hub.docker.com/r/matrixdotorg/synapse/tags?page=1&ordering=last_updated), I believe that that workflow both tests docker builds as well as publishes them (if the branch has an associated release tag).

This isn't necessary in Sygnal, as we already have a workflow for testing docker builds.

File: .github/workflows/docker_push.yml (hunk 3 above)
Comment (anoadragon453) [reply .github/workflows/docker_push.yml:40]: yeah, I'd cut it.

**Review by anoadragon453 — DISMISSED** [review]
lgtm once the above is sorted.

**Comment by bradtgmurray** [comment]
Now that this is resolved, can we pretty please get a point release tagged and pushed? Looking forward to trying out the db-less sygnal release, but unfortunately the container isn't available on dockerhub. https://hub.docker.com/r/matrixdotorg/sygnal/tags

**Comment by reivilibre** [comment]
> Now that this is resolved, can we pretty please get a point release tagged and pushed? Looking forward to trying out the db-less sygnal release, but unfortunately the container isn't available on dockerhub. https://hub.docker.com/r/matrixdotorg/sygnal/tags

That's the plan, yup, sorry for taking my sweet time with it :)

**Comment by reivilibre** [comment]
@bradtgmurray 0.10.1 released -- go and have fun! :D

**Pull request merged**